cpp/src/community/louvain.cuh

```diff
@@ -0,0 +1,639 @@
+      indices_v_(graph.indices, graph.indices + graph.number_of_edges),
+      weights_v_(graph.edge_data, graph.edge_data + graph.number_of_edges),
+      src_indices_v_(graph.number_of_edges),
+      vertex_weights_v_(graph.number_of_vertices),
+      cluster_weights_v_(graph.number_of_vertices),
+      cluster_v_(graph.number_of_vertices),
+      tmp_arr_v_(graph.number_of_vertices),
+      cluster_inverse_v_(graph.number_of_vertices),
+      number_of_vertices_(graph.number_of_vertices),
+      number_of_edges_(graph.number_of_edges),
+      stream_(stream)
+  {
+  }
+
+  weight_t modularity(weight_t total_edge_weight,
+                      weight_t resolution,
+                      graph_t const &graph,
+                      vertex_t const *d_cluster)
+  {
+    vertex_t n_verts = graph.number_of_vertices;
+
+    rmm::device_vector<weight_t> inc(n_verts, weight_t{0.0});
+    rmm::device_vector<weight_t> deg(n_verts, weight_t{0.0});
+
+    edge_t const *d_offsets   = graph.offsets;
+    vertex_t const *d_indices = graph.indices;
+    weight_t const *d_weights = graph.edge_data;
+    weight_t *d_inc           = inc.data().get();
+    weight_t *d_deg           = deg.data().get();
+
+    // FIXME:  Already have weighted degree computed in main loop,
+    //         could pass that in rather than computing d_deg... which
+    //         would save an atomicAdd (synchronization)
+    //
+    thrust::for_each(
+      rmm::exec_policy(stream_)->on(stream_),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(graph.number_of_vertices),
+      [d_inc, d_deg, d_offsets, d_indices, d_weights, d_cluster] __device__(vertex_t v) {
+        vertex_t community = d_cluster[v];
+        weight_t increase{0.0};
+        weight_t degree{0.0};
+
+        for (edge_t loc = d_offsets[v]; loc < d_offsets[v + 1]; ++loc) {
+          vertex_t neighbor = d_indices[loc];
+          degree += d_weights[loc];
+          if (d_cluster[neighbor] == community) { increase += d_weights[loc]; }
+        }
+
+        if (degree > weight_t{0.0}) atomicAdd(d_deg + community, degree);
+        if (increase > weight_t{0.0}) atomicAdd(d_inc + community, increase);
+      });
+
+    weight_t Q = thrust::transform_reduce(
+      rmm::exec_policy(stream_)->on(stream_),
+      thrust::make_counting_iterator(0),
+      thrust::make_counting_iterator(graph.number_of_vertices),
+      [d_deg, d_inc, total_edge_weight, resolution] __device__(vertex_t community) {
+        return ((d_inc[community] / total_edge_weight) - resolution *
+                                                           (d_deg[community] * d_deg[community]) /
+                                                           (total_edge_weight * total_edge_weight));
+      },
+      weight_t{0.0},
+      thrust::plus<weight_t>());
+
+    return Q;
+  }
+
+  virtual std::pair<int, weight_t> compute(vertex_t *d_cluster_vec,
```

Comment:
See comment on using a functor (hence, overloading `operator()(...)`)

Reply:
Done in next commit